add `--pass-through` to build and repl

Profpatsch opened the pull request:
It should be possible to pass through arguments to the called `purs` commands,
like in this case `compile` and `repl`.
This is needed for e.g. using `purs compile --json-errors` to integrate
psc-package into editors.

Profpatsch committed:
add `--pass-through` to build and repl

It should be possible to pass through arguments to the called `purs` commands,
like in this case `compile` and `repl`.
This is needed for e.g. using `purs compile --json-errors` to integrate
psc-package into editors.

paf31 commented:
Is there any way to have pass-through args be separated from regular ones by just `--`, which I think is usual?

Thanks for the PR!

Profpatsch commented:
You mean `psc-package build <args> -- <passthrough-args>`? Afaik `--` means there won’t be any options afterwards, everything should be interpreted as filename (to prevent files starting with `--` to be interpreted as options, used by globbing attacks)

I tried it out in `vscode` and manually, `--pass-through="--option" --pass-through="--other-option"` works like a charm.

Pauan commented:
Both `yarn` and `npm` use `--` to mean "pass through the remaining arguments"

e.g. you can use `yarn run -- foo -bar` or `npm run foo -- -bar`

I think this is a very logical way of handling it, because `--` literally means "don't interpret the remaining arguments as options", so naturally they will be passed through unchanged.

I've never heard "interpret everything afterwards as a filename". The [POSIX specification](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html) simply says:

> The first `--` argument that is not an option-argument should be accepted as a delimiter indicating the end of options. Any following arguments should be treated as operands, even if they begin with the `-` character.

It is up to the program to decide *how* it treats arguments. The `--` simply specifies where the options end, nothing more.

Not every program treats arguments as filenames, so requiring programs to treat everything after `--` as a filename (even if they don't normally treat arguments as filenames) would be very odd.

Profpatsch commented:
> Not every program treats arguments as filenames, so requiring programs to treat everything after -- as a filename (even if they don't normally treat arguments as filenames) would be very odd.

Can we be sure none of these commands wants files after `--`? Because when that happens, it will be very ad-hoc to keep pass-through and files separate.

Pauan commented:
@Profpatsch If you want to pass a filename which starts with `-` (for whatever reason), you can do that by using two `--`, like this:

```
psc-package build -- -purs-options-go-here -- -filename
```

The first `--` is an "escape" for `psc-package`, the second `--` is an "escape" for `purs`

This should work because everything after `--` is not treated as an option, including `--` itself. So the second `--` gets passed through to `purs`

Profpatsch commented:
Okay, this is just silly now. I’m definitely against using `--`, because it introduces special cases and
`--pass-trough` is semantically obvious.

Pauan commented:
In my opinion, your choices are:

1. Support `--` and `--pass-through`

2. Support `--`

I don't care much one way or the other whether `--pass-through` is supported, but `--` must be supported for POSIX compatibility (and familiarity with existing tools like npm and yarn).

Pauan commented:
Oh yes, I forgot, pulp also uses `--` to separate `pulp` options from `purs` options: https://github.com/bodil/pulp/issues/220

https://github.com/bodil/pulp#pass-through-options

So it's not just about POSIX compatibility and familiarity with JS tools, it's also about familiarity with PureScript tools as well.

kritzcreek commented:
I actually really like that we separate dependency management from build tooling in PureScript. I don't think `psc-package` should support a bunch of build options. The build command is a stop gap until our build tools decide to support `psc-package`. It even says so in the README:

> This command is provided as a convenience until external tools add support for psc-package. It might be removed in future.

If the current form of the `build` command isn't enough it's easy to write a little script instead (it only supplies globs anyway).

Profpatsch commented:
> The build command is a stop gap until our build tools decide to support psc-package.

I’m not quite sure what that means for this PR; do you think `purs` should know about `psc-package`? That sounds like unnecessary entanglement to me. Maybe I misunderstood and you mean it shouldn’t be possible to give `build` flags for the underlying builder (in this case `purs compile`). But then I don’t know how it can be done (without being unable to tell `purs` stuff like “give me json output”; though in this case it could be added as a `psc-package` option—but then we’d have to standardize the JSON format that is outputted).

kritzcreek commented:
> do you think purs should know about psc-package?

No. `pulp` is the build tool I'm primarily talking about, but I know there are different solutions using shake or make

> Maybe I misunderstood and you mean it shouldn’t be possible to give build flags for the underlying builder (in this case purs compile).

I don't think there should be an "underlying builder" at all. A build tool would collect the output of `psc-package sources` and pass it on to `purs compile` (That's the part I said is pretty easy to script right now, actually you don't even need `psc-package sources` just a simple `purs compile "./.psc-package/psc-0.11.5/*/*/src/**/*.purs" "src/**/*.purs"` does the job).

> without being unable to tell purs stuff like “give me json output”; though in this case it could be added as a psc-package option—but then we’d have to standardize the JSON format that is outputted

That's exactly what happens when you pass `--json-errors` to `purs compile`.

Profpatsch commented:
Okay, I’m starting to understand the intended structure:

`purs compile` should handle all build logic, provided it’s given a set of all necessary modules.
`psc-package` should only care about purescript-only packages, resolve the versions and put them somewhere our build tool can find them.
The build tool, for example `pulp`, is a wrapper that provides everything else and correctly hands over files to `purs compile`.

One open question would be how to handle “native” JS dependencies (probably coming from npm).

I suggest merging this PR as a temporary measure, and opening a discussion issue on how to proceed with deprecating `build` (and `repl`?) for `psc-package`.

paf31 commented:
Can't we implement `--`? I tried this in the REPL:

```
Options.Applicative> execParserPure (prefs mempty) (info ((,) <$> switch (long "flag") <*> many (strArgument mempty)) mempty) ["--flag", "--", "--json-errors"]
Success (True,["--json-errors"])
```

Here `--flag` is an argument to `psc-package`, and `--json-errors` is an argument to `purs`.

Profpatsch committed:
switch from `--pass-through` to implicit passthrough

People think this is a more usable solution. I disagree, but it’s not really
important enough to argue over.

Profpatsch commented:
Done.

paf31 commented:
Thanks!
